src/sentry/api/endpoints/user_emails.py

```diff
@@ -0,0 +1,240 @@
+from __future__ import absolute_import
+
+import logging
+
+from django.db import IntegrityError, transaction
+from django.db.models import Q
+from rest_framework import serializers, status
+from rest_framework.response import Response
+
+from sentry import newsletter
+from sentry.api.bases.user import UserEndpoint
+from sentry.api.decorators import sudo_required
+from sentry.api.serializers import serialize
+from sentry.models import User, UserEmail, UserOption
+
+logger = logging.getLogger('sentry.accounts')
+
+
+InvalidEmailResponse = Response({'detail': 'Invalid email'}, status=status.HTTP_400_BAD_REQUEST)
+
+
+class InvalidEmailError(Exception):
+    pass
+
+
+class DuplicateEmailError(Exception):
+    pass
+
+
+class EmailSerializer(serializers.Serializer):
+    email = serializers.EmailField(required=False)
+
+
+def get_email(request):
+    """
+    Returns None if invalid email else returns a normalized email (lowercase + whitespace strip)
+    """
+
+    serializer = EmailSerializer(data=request.DATA)
+
+    # Bad email
+    if not serializer.is_valid():
+        return None
+
+    return serializer.object['email'].lower().strip()
+
+
+def add_email(email, user):
+    """
+    Adds an email to user account
+
+    Can be either primary or secondary
+    """
+
+    # Bad email
+    if email is None:
+        raise InvalidEmailError
+
+    # check if this email already exists for user
+    if email and UserEmail.objects.filter(
```

Comment:
do you need this `if email` if you've previously returned when `email is None`?

Reply:
on second thought, you may not even need this block. i don't really think there's any harm in just trying to create the email and returning on an integrity error (which you do below)